backend/api/experiment.proto

```diff
@@ -150,4 +164,19 @@ message Experiment {
   // Optional input field. Specify which resource this run belongs to.
   // For Experiment, the only valid resource reference is a single Namespace.
   repeated ResourceReference resource_references = 5;
+
+  enum StorageState {
```

Comment:
Just out of curiosity, does it make sense to have the default value as `STORAGESTATE_UNSPECIFIED = 0;`?

Reply:
I actually want to do that! I have been debating with myself since RUN's storage _state has only two enum values (archived and unarchived). But I actually want to have one more enum value like RESERVED or UNSPECIFIED as you mentioned. The reason is that I found the first enum value being 0 will cause some wierd behavior when we return the api type; and thus I want to have archive and unarchive take non-0 values (and i.e., has RESERVED or UNSPECIFIED taking 0 value)

Reply:
Nice to know the omitempty in json output! The case in StorageState might be slightly different here. If boolean field is omitted from a struct, when user try to read the field, the default value of boolean filed when not set is 0 and 0 will be returned to the user, and since the default 0 happens to be the same as the omitted value, user actually gets the correct value. And I feel that is the reason why json output can omit some values safely.

However, StorageState is enum strings. It only takes two strings "storagestate_archived" and "storagestate_available". e.g., in run api test, we expect https://github.com/kubeflow/pipelines/blob/d9bb1be25b838142267acbf83ca8f1a5bf33ef54/backend/test/integration/run_api_test.go#L202
Meanwhile,  if a string typed field got omitted, the value will become "". In other words, a generic string typed field can be omitted, but a enum string type can't be omitted. And this is the symptom I ran into. The field disappear and the string value returned to me becomes "".  This is not one of the two enum string values.

In api converter, we convert model type to api type. In model type, storage state is indeed enum integer, and I verified that the model type is actually retrieved correctly as being 0. We there use that 0 to find the "storagestate_available" string. But from that point on, the string seems missing from the returned struct. I'll take another look and keep your note about "omitempty" in mind to investigate further.

Reply:
Nice to know the json omitempty behavior! My understanding is that boolean and generic string typed fields can be omitted because the default value (used when field not explicitly set) happens to be the same as the omitted value and hence the value is correct. Our StorageState type is a enum string which takes two values, e.g., taking Run as an example, https://github.com/kubeflow/pipelines/blob/4c491823a0f169db6eec5bb3415e1ca31dcf9ee5/backend/api/go_http_client/run_model/run_storage_state.go#L33
I don't think enum string typed field can be marked omitempty. Something else might be happening here...

Reply:
SGTM if you decides. Just wondering if you can make STORAGESTATE_AVAILABLE the default value for backfilled DB data?

Reply:
I chose not to do the data backfill. While manually testing this implementation on a KFP deployment, the existing rows seem to have "" populated in the new column (storage state). So when we retrieve the rows in experiment_store.go, I put a comment on the treatment of "", basically explaining that why we ran into "" in this column (while we shouldn't in a normal case) and how we patch for this legacy value.

Reply:
If we are still dealing with "" in our code, what has been improved by adding an extra enum value?
Sounds to me keeping the enum as 2 values and deal with "" in clients have fewer changes needed.

Reply:
The column got a "" value as that's the default for type string. I wonder what was the thought behind storing as string type vs int type (or even tinyint). As far as I can think of, string only benefits readability if one looks directly into the DB, whereas int type can save storage, and it defaults to 0, which will naturally match whatever the default enum value is.

Reply:
First of all, both string and int types seem fine to me. I don't think this is a super critical decision (or life-and-death decision) in our case. E.g., even considering size, our strings are fixed length (since only predefined strings are allowed) and thus not a huge deal. Then why choosing string? To be consistent with Run storage schema. So the later developer might get better code readability (note: not readability of data in db). They'll naturally get the hint that storage state in Run and Experiment are meant to express similar notation. In other words, if Run has this field as int in the first place, int can b used here as well. Again, IMHO, not a crucial enough decision either in terms of performance or in terms of code/data readability, and arguments can be made on either side .

Reply:
Yeah, I agree it's not a super critical decision. And as you mentioned, consistency with existing Run state might be a good reason to stay with string.